File: 205.isomorphic-strings.rs
/*
 * @lc app=leetcode id=205 lang=rust
 *
 * [205] Isomorphic Strings
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn is_isomorphic(s: String, t: String) -> bool {
        let l1 = s.len();
        let l2 = t.len();
        if l1 != l2 {
            return false;
        }

        let mut m1: HashMap<char, usize> = HashMap::new();
        let mut m2: HashMap<char, usize> = HashMap::new();

        let s = &mut s.chars();
        let t = &mut t.chars();

        for i in 0..l1 {
            let ch1 = s.next().unwrap();
            let ch2 = t.next().unwrap();

            if !m1.insert(ch1, i).eq(&m2.insert(ch2, i)){
              return false;
            }
        }

        true
    }
}
// @lc code=end

File: 205.isomorphic-strings.cpp
/*
 * @lc app=leetcode id=205 lang=cpp
 *
 * [205] Isomorphic Strings
 */

// @lc code=start
class Solution {
public:
    // bool isIsomorphic(string s, string t) {
    bool isIsomorphic(const string& s, const string& t) {
        int l1 = s.size();
        int l2 = t.size();
        if(l1 != l2) return false;

        unordered_map<char, char> m1;
        unordered_map<char, char> m2;

        for(int i = 0; i < l1; ++i) {
            char c1 = s[i];
            char c2 = t[i];

            if(m1.find(c1) == m1.end() && m2.find(c2) == m2.end()) {
                m1[c1] = c2;
                m2[c2] = c1;
                continue;
            }

            if(m1[c1] != c2 || m2[c2] != c1) return false;
        }

        return true;
    }
};
// @lc code=end

